network/write.go

```diff
@@ -0,0 +1,157 @@
+
+	if err != nil {
+		return nil, err
+	}
+	return &write{
+		client: httpClient,
+		cfg:    cc,
+		log:    log.With(l, "name", "loop", "url", cc.URL),
+		stats:  statsResult,
+	}, nil
+}
+
+// trySend is the core functionality for sending data to an endpoint. It will attempt retries as defined in MaxRetryAttempts.
+func (l *write) trySend(buf []byte, ctx context.Context) {
+	attempts := 0
+	for {
+		result := l.send(buf, ctx, attempts)
+
+		if result.err != nil {
+			level.Error(l.log).Log("msg", "error in sending telemetry", "err", result.err.Error())
+		}
+		if result.successful {
+			return
+		}
+		if !result.recoverableError {
+			return
+		}
+		attempts++
+		if attempts > int(l.cfg.MaxRetryAttempts) && l.cfg.MaxRetryAttempts > 0 {
+			level.Debug(l.log).Log("msg", "max retry attempts reached", "attempts", attempts)
+			return
+		}
+		// Sleep between attempts.
+		time.Sleep(result.retryAfter)
+	}
+
+}
+
+type sendResult struct {
+	err              error
+	successful       bool
+	recoverableError bool
+	retryAfter       time.Duration
+	statusCode       int
+	networkError     bool
+	duration         time.Duration
+}
+
+// send is the main work loop of the loop.
+func (l *write) send(buf []byte, ctx context.Context, retryCount int) sendResult {
+	start := time.Now()
+	result := sendResult{}
+	defer func() {
+		result.duration = time.Since(start)
+		l.stats(result)
+	}()
+	httpReq, err := http.NewRequest("POST", l.cfg.URL, bytes.NewReader(buf))
+	if err != nil {
+		result.err = err
+		result.recoverableError = true
+		result.networkError = true
+		return result
+	}
+	httpReq.Header.Add("Content-Encoding", "snappy")
+	httpReq.Header.Set("Content-Type", "application/x-protobuf")
+	httpReq.Header.Set("User-Agent", l.cfg.UserAgent)
+	httpReq.Header.Set("X-Prometheus-Remote-Write-Version", "0.1.0")
+	if l.cfg.BasicAuth != nil {
+		httpReq.SetBasicAuth(l.cfg.BasicAuth.Username, l.cfg.BasicAuth.Password)
+	} else if l.cfg.BearerToken != "" {
+		httpReq.Header.Set("Authorization", "Bearer "+string(l.cfg.BearerToken))
+	}
+
+	if retryCount > 0 {
+		httpReq.Header.Set("Retry-Attempt", strconv.Itoa(retryCount))
+	}
+	ctx, cncl := context.WithTimeout(ctx, l.cfg.Timeout)
+	defer cncl()
+	resp, err := l.client.Do(httpReq.WithContext(ctx))
+	// Network errors are recoverable.
+	if err != nil {
+		result.err = err
+		result.networkError = true
+		result.recoverableError = true
+		result.retryAfter = l.cfg.RetryBackoff
+		return result
+	}
+	result.statusCode = resp.StatusCode
+	defer resp.Body.Close()
+	// 500 errors are considered recoverable.
+	if resp.StatusCode/100 == 5 || resp.StatusCode == http.StatusTooManyRequests {
+		result.err = fmt.Errorf("server responded with status code %d", resp.StatusCode)
+		result.retryAfter = retryAfterDuration(l.cfg.RetryBackoff, resp.Header.Get("Retry-After"))
+		result.recoverableError = true
+		return result
+	}
+	// Status Codes that are not 500 or 200 are not recoverable and dropped.
```

Comment:
Comment implies 500s should be treated in a special way, but then that doesn't appear to be represented in the code.

Reply:
Its that they are considered recoverable. write.go:60 has the check, ie they will be retired.

```
		if !result.recoverableError {
			return
		}
```